convex/lib/rowLevelSecurity.ts

```diff
@@ -0,0 +1,404 @@
+import {
+  DatabaseReader,
+  DatabaseWriter,
+  DocumentByInfo,
+  DocumentByName,
+  Expression,
+  FilterBuilder,
+  FunctionArgs,
+  GenericDataModel,
+  GenericTableInfo,
+  IndexRange,
+  IndexRangeBuilder,
+  Indexes,
+  MutationCtx,
+  NamedIndex,
+  NamedSearchIndex,
+  NamedTableInfo,
+  OrderedQuery,
+  PaginationOptions,
+  PaginationResult,
+  Query,
+  QueryCtx,
+  QueryInitializer,
+  SearchFilter,
+  SearchFilterBuilder,
+  SearchIndexes,
+  TableNamesInDataModel,
+} from "convex/server";
+import { GenericId } from "convex/values";
+
+export type Rule<
+  Ctx,
+  DataModel extends GenericDataModel,
+  TableName extends TableNamesInDataModel<DataModel>
+> = (
+  ctx: Ctx,
+  message: DocumentByName<DataModel, TableName>
+) => Promise<boolean>;
+
+export type Rules<Ctx, DataModel extends GenericDataModel> = {
+  [T in TableNamesInDataModel<DataModel>]?: Rule<Ctx, DataModel, T>;
+};
+
+/**
+ * Apply row level security (RLS) to queries and mutations with the returned
+ * middleware functions.
+ *
+ * Example:
+ * ```
+ * // Defined in a common file so it can be used by all queries and mutations.
+ * const {withMutationRLS} = RowLevelSecurity<{auth: Auth, db: DatabaseReader}, DataModel>(
+ *  {
+ *    cookies: async ({auth}, cookie) => !cookie.eaten,
+ *  },
+ *  {
+ *    cookies: async ({auth, db}, cookie) => {
+ *      const user = await getUser(auth, db);
+ *      return user.isParent;  // only parents can reach the cookies.
+ *    }
+ *  }
+ * );
+ * // Mutation with row level security enabled.
+ * export const eatCookie = mutation(withMutationRLS(
+ *  async ({db}, {cookieId}) => {
+ *   // throws "does not exist" error if cookie is already eaten or doesn't exist.
+ *   // throws "write access" error if authorized user is not a parent.
+ *   await db.patch(cookieId, {eaten: true});
+ * }));
+ * ```
+ *
+ * Notes:
+ * * Rules may read any row in `db` -- rules do not apply recursively within the
+ *   rule functions themselves.
+ * * Tables with no rule default to full access.
+ * * Middleware functions like `withUser` can be composed with RowLevelSecurity
+ *   to cache fetches in `ctx`. e.g.
+ * ```
+ * const {withQueryRLS} = RowLevelSecurity<{user: Doc<"users">}, DataModel>(
+ *  {
+ *    cookies: async ({user}, cookie) => user.isParent,
+ *  }
+ * );
+ * export default query(withUser(withRLS(...)));
+ * ```
+ *
+ * @param readAccessRules - rule for each table, determining whether a row
+ *  should be visible.
+ * @param writeAccessRules - rule for each table, determining whether a row
+ *  may be written by `patch`, `replace`, or `delete`.
+ *  Rules do not restrict `db.insert`.
```

Comment:
so any user can create rows? can't we pass the new document in? Add _creationTime as Date.now(), and _id from the insert? Let's make sure to call this out as a caveat in the readme. in the future we could also extend like:
```
{
  cookies: {
    read: ...
    write: ...
    validateInserts: true
  },
  ...
}
```
where your validation function will take in the null on inserts, if validating the new object doesn't make sense.

Reply:
or any made up _id - ideally the auth check shouldn't bleed info out, and the _id doesn't exist yet so they don't have much to check. or we pass in the `withoutSystemFields` type and hope they don't need the _id? Or make _id optional so they have to check in their authorization fn, and `null` means it hasn't been inserted yet

Reply:
how about a separate validation rule for inserts that can take in a `WithoutSystemFields`

Reply:
how about read / write / modify, where modify = "write" and write = "insert"?
"write" seems more like insert. or just change "write" to "modify" and leave insert?